File: exact-model-counter-ab6fb0/solution/counter.cpp
/* Exact Model Counter for MC Competition DIMACS format.
 * Compiled with: g++ -O2 -Wall -std=c++17 -o mc mc.cpp -lgmpxx -lgmp
 *
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <set>
#include <map>
#include <unordered_map>
#include <algorithm>
#include <functional>
#include <cstdlib>
#include <gmpxx.h>

using namespace std;

/* ---------- weight parsing ---------- */

mpq_class parse_decimal(const string& s) {
    size_t dot = s.find('.');
    if (dot == string::npos) return mpq_class(s);

    string ip = s.substr(0, dot);
    string fp = s.substr(dot + 1);
    if (ip.empty()) ip = "0";

    bool neg = (!ip.empty() && ip[0] == '-');
    string aip = neg ? ip.substr(1) : ip;
    if (aip.empty()) aip = "0";

    mpz_class num(aip + fp);
    if (neg) num = -num;
    mpz_class den;
    mpz_ui_pow_ui(den.get_mpz_t(), 10, fp.length());
    mpq_class r(num, den);
    r.canonicalize();
    return r;
}

mpq_class parse_weight(const string& s) {
    /* fraction: P/Q */
    size_t sl = s.find('/');
    if (sl != string::npos) {
        mpz_class num(s.substr(0, sl));
        mpz_class den(s.substr(sl + 1));
        mpq_class r(num, den);
        r.canonicalize();
        return r;
    }

    /* scientific: XeY or XEY */
    size_t ep = s.find('e');
    if (ep == string::npos) ep = s.find('E');
    if (ep != string::npos) {
        mpq_class mant = parse_decimal(s.substr(0, ep));
        int exp = stoi(s.substr(ep + 1));
        mpz_class tp;
        mpz_ui_pow_ui(tp.get_mpz_t(), 10, abs(exp));
        if (exp >= 0) mant *= mpq_class(tp);
        else          mant /= mpq_class(tp);
        mant.canonicalize();
        return mant;
    }

    return parse_decimal(s);
}

/* ---------- DIMACS parser ---------- */

struct Instance {
    int nvars = 0, nclauses = 0;
    bool weighted = false;
    vector<vector<int>> clauses;
    map<int, mpq_class> weights;
};

Instance parse_dimacs(const string& fname) {
    Instance inst;
    ifstream f(fname);
    if (!f) { cerr << "Cannot open " << fname << "\n"; exit(1); }

    string line;
    while (getline(f, line)) {
        if (line.empty()) continue;
        if (line[0] == 'c') {
            istringstream iss(line);
            string c_tok; iss >> c_tok;
            string tok;
            if (!(iss >> tok)) continue;
            if (tok == "t") {
                string track; iss >> track;
                if (track == "wmc") inst.weighted = true;
            } else if (tok == "p") {
                string dir; iss >> dir;
                if (dir == "weight") {
                    int lit; string ws;
                    if (iss >> lit >> ws)
                        inst.weights[lit] = parse_weight(ws);
                }
            }
        } else if (line[0] == 'p') {
            istringstream iss(line);
            string p, cnf;
            iss >> p >> cnf >> inst.nvars >> inst.nclauses;
        } else {
            istringstream iss(line);
            vector<int> cl;
            int l;
            while (iss >> l && l != 0) cl.push_back(l);
            if (!cl.empty()) inst.clauses.push_back(cl);
        }
    }
    return inst;
}

/* ---------- model counter ---------- */

class Counter {
    Instance& inst;
    unordered_map<string, mpq_class> cache;

    mpq_class w(int lit) {
        if (!inst.weighted) return mpq_class(1);
        auto it = inst.weights.find(lit);
        return it != inst.weights.end() ? it->second : mpq_class(1);
    }

    mpq_class free_w(int var) {
        return inst.weighted ? w(var) + w(-var) : mpq_class(2);
    }

    /* Simplify clauses by assigning var=val.
       Returns {new_clauses, conflict}. */
    pair<vector<vector<int>>, bool> assign(
            const vector<vector<int>>& cls, int var, bool val) {
        int tl = val ? var : -var;
        int fl = val ? -var : var;
        vector<vector<int>> out;
        for (auto& c : cls) {
            bool sat = false;
            vector<int> nc;
            for (int l : c) {
                if (l == tl) { sat = true; break; }
                if (l != fl) nc.push_back(l);
            }
            if (sat) continue;
            if (nc.empty()) return {{}, true};
            out.push_back(nc);
        }
        return {out, false};
    }

    /* Unit propagation.  Returns {factor, clauses, vars, conflict}. */
    struct UP {
        mpq_class factor;
        vector<vector<int>> cls;
        set<int> vars;
        bool conflict;
    };

    UP propagate(vector<vector<int>> cls, set<int> vars) {
        UP r;
        r.factor = 1;
        r.conflict = false;
        bool go = true;
        while (go) {
            go = false;
            for (auto& c : cls) {
                if (c.size() == 1) {
                    int lit = c[0], v = abs(lit);
                    r.factor *= w(lit);
                    vars.erase(v);
                    auto [nc, conf] = assign(cls, v, lit > 0);
                    if (conf) { r.conflict = true; return r; }
                    cls = nc;
                    go = true;
                    break;
                }
            }
        }
        r.cls = cls;
        r.vars = vars;
        return r;
    }

    /* Connected components via union-find. */
    vector<pair<vector<vector<int>>, set<int>>> components(
            const vector<vector<int>>& cls) {
        map<int, int> par;
        set<int> cv;
        for (auto& c : cls)
            for (int l : c) { int v = abs(l); cv.insert(v); par[v] = v; }

        function<int(int)> find = [&](int x) -> int {
            return par[x] == x ? x : par[x] = find(par[x]);
        };

        for (auto& c : cls) {
            int f = abs(c[0]);
            for (size_t i = 1; i < c.size(); i++) {
                int a = find(f), b = find(abs(c[i]));
                if (a != b) par[a] = b;
            }
        }

        map<int, vector<vector<int>>> cc;
        map<int, set<int>> ccv;
        for (auto& c : cls) {
            int root = find(abs(c[0]));
            cc[root].push_back(c);
            for (int l : c) ccv[root].insert(abs(l));
        }

        vector<pair<vector<vector<int>>, set<int>>> out;
        for (auto& [root, c] : cc)
            out.push_back({c, ccv[root]});
        return out;
    }

    string cache_key(const vector<vector<int>>& cls) {
        auto sc = cls;
        for (auto& c : sc) sort(c.begin(), c.end());
        sort(sc.begin(), sc.end());
        string k;
        for (auto& c : sc) {
            for (int l : c) { k += to_string(l); k += ','; }
            k += ';';
        }
        return k;
    }

public:
    Counter(Instance& i) : inst(i) {}

    mpq_class count(vector<vector<int>> cls, set<int> vars) {
        /* unit propagation */
        auto up = propagate(cls, vars);
        if (up.conflict) return mpq_class(0);
        cls = up.cls;
        vars = up.vars;

        /* all clauses satisfied */
        if (cls.empty()) {
            mpq_class r = up.factor;
            for (int v : vars) r *= free_w(v);
            return r;
        }

        /* identify clause variables and free variables */
        set<int> cv;
        for (auto& c : cls) for (int l : c) cv.insert(abs(l));
        mpq_class ff(1);
        for (int v : vars)
            if (!cv.count(v)) ff *= free_w(v);

        /* component decomposition */
        auto comps = components(cls);
        if (comps.size() > 1) {
            mpq_class r = up.factor * ff;
            for (auto& [cc, ccv] : comps)
                r *= count(cc, ccv);
            return r;
        }

        /* cache lookup */
        string key = cache_key(cls);
        auto it = cache.find(key);
        if (it != cache.end())
            return up.factor * ff * it->second;

        /* branch on most-frequent variable */
        map<int, int> freq;
        for (auto& c : cls) for (int l : c) freq[abs(l)]++;
        int bv = 0, bf = 0;
        for (auto& [v, f] : freq)
            if (f > bf) { bf = f; bv = v; }

        set<int> inner = cv;
        inner.erase(bv);

        mpq_class ct(0), cf(0);
        {
            auto [tc, tconf] = assign(cls, bv, true);
            if (!tconf) ct = w(bv) * count(tc, inner);
        }
        {
            auto [fc, fconf] = assign(cls, bv, false);
            if (!fconf) cf = w(-bv) * count(fc, inner);
        }

        mpq_class comp = ct + cf;
        cache[key] = comp;
        return up.factor * ff * comp;
    }

    pair<bool, mpq_class> solve() {
        set<int> all;
        for (int i = 1; i <= inst.nvars; i++) all.insert(i);
        mpq_class r = count(inst.clauses, all);
        return {r != 0, r};
    }
};

/* ---------- main ---------- */

int main(int argc, char** argv) {
    if (argc != 2) {
        cerr << "Usage: mc <file.cnf>\n";
        return 1;
    }
    Instance inst = parse_dimacs(argv[1]);
    Counter c(inst);
    auto [sat, cnt] = c.solve();

    cout << (sat ? "s SATISFIABLE" : "s UNSATISFIABLE") << "\n";
    if (inst.weighted) {
        cout << "c s type wmc\n";
        cnt.canonicalize();
        cout << "c s exact rational "
             << cnt.get_num() << "/" << cnt.get_den() << "\n";
    } else {
        cout << "c s type mc\n";
        cout << "c s exact arb int " << cnt.get_num() << "\n";
    }
    return 0;
}

File: exact-model-counter-ab6fb0/solution/counter.py
#!/usr/bin/env python3
"""
Exact model counter for MC competition DIMACS format.
Supports mc (unweighted) and wmc (weighted) tracks.
Uses DPLL search with unit propagation, component decomposition, and memoization.

"""

import sys
import math
from fractions import Fraction
from decimal import Decimal
from collections import defaultdict


def parse_weight(s):
    """Parse a weight string (decimal, fraction, or scientific) into exact Fraction."""
    if '/' in s:
        num, den = s.split('/')
        return Fraction(int(num), int(den))
    return Fraction(Decimal(s))


def parse_dimacs(filename):
    """Parse DIMACS CNF with MC competition directives."""
    clauses = []
    num_vars = 0
    mode = 'mc'
    weights = {}

    with open(filename) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            tokens = line.split()
            if tokens[0] == 'c':
                if len(tokens) >= 3 and tokens[1] == 't':
                    mode = tokens[2]
                elif (len(tokens) >= 6 and tokens[1] == 'p'
                      and tokens[2] == 'weight'):
                    lit = int(tokens[3])
                    weights[lit] = parse_weight(tokens[4])
            elif tokens[0] == 'p':
                num_vars = int(tokens[2])
            else:
                lits = [int(x) for x in tokens]
                if lits and lits[-1] == 0:
                    lits = lits[:-1]
                if lits:
                    clauses.append(tuple(lits))

    return num_vars, clauses, mode, weights


class ModelCounter:
    """DPLL-based exact model counter with component caching."""

    def __init__(self, num_vars, clauses, mode, weights):
        self.num_vars = num_vars
        self.clauses = clauses
        self.mode = mode
        self.weights = weights
        self.is_weighted = mode in ('wmc', 'pwmc', 'wpmc')
        self.cache = {}

    def lit_weight(self, lit):
        """Get weight of a literal."""
        if not self.is_weighted:
            return Fraction(1)
        return self.weights.get(lit, Fraction(1))

    def free_weight(self, var):
        """Get the combined weight of both polarities of a free variable."""
        return self.lit_weight(var) + self.lit_weight(-var)

    def simplify(self, clauses, var, val):
        """Simplify clause list after assigning var=val.
        Returns simplified clause list, or None on conflict (empty clause)."""
        result = []
        for clause in clauses:
            satisfied = False
            remaining = []
            for lit in clause:
                v = abs(lit)
                if v == var:
                    if (lit > 0) == val:
                        satisfied = True
                        break
                    # literal is false under this assignment, skip it
                else:
                    remaining.append(lit)
            if satisfied:
                continue
            if not remaining:
                return None  # empty clause = conflict
            result.append(tuple(remaining))
        return result

    def unit_propagate(self, clauses, unassigned):
        """Iteratively assign unit clauses.
        Returns (clauses, unassigned, accumulated_weight, conflict_flag)."""
        weight = Fraction(1)
        unass = set(unassigned)
        cls = list(clauses)

        changed = True
        while changed:
            changed = False
            for clause in cls:
                if len(clause) == 1:
                    lit = clause[0]
                    var = abs(lit)
                    val = lit > 0
                    weight *= self.lit_weight(lit)
                    unass.discard(var)
                    cls = self.simplify(cls, var, val)
                    if cls is None:
                        return None, None, None, True
                    changed = True
                    break  # restart scan after modification

        return cls, unass, weight, False

    def find_components(self, clauses, unassigned):
        """Group clauses into connected components by shared variables."""
        if not clauses:
            return []

        clause_vars = []
        for clause in clauses:
            cvars = frozenset(abs(lit) for lit in clause if abs(lit) in unassigned)
            clause_vars.append(cvars)

        var_to_cis = defaultdict(set)
        for i, cvs in enumerate(clause_vars):
            for v in cvs:
                var_to_cis[v].add(i)

        visited = set()
        components = []
        for start in range(len(clauses)):
            if start in visited:
                continue
            comp_cis = set()
            stack = [start]
            while stack:
                ci = stack.pop()
                if ci in visited:
                    continue
                visited.add(ci)
                comp_cis.add(ci)
                for v in clause_vars[ci]:
                    for adj_ci in var_to_cis[v]:
                        if adj_ci not in visited:
                            stack.append(adj_ci)

            comp_cls = [clauses[i] for i in comp_cis]
            comp_vs = set()
            for i in comp_cis:
                comp_vs |= clause_vars[i]
            components.append((comp_vs, comp_cls))

        return components

    def _free_count(self, free_vars):
        """Compute the count contribution of free (unconstrained) variables."""
        if not free_vars:
            return Fraction(1)
        if self.is_weighted:
            result = Fraction(1)
            for v in free_vars:
                result *= self.free_weight(v)
            return result
        return Fraction(2) ** len(free_vars)

    def count(self, clauses, unassigned):
        """Recursively count models with caching and component decomposition."""
        # Base: no clauses left
        if not clauses:
            return self._free_count(unassigned)

        # Unit propagation
        up_cls, up_unass, up_weight, conflict = self.unit_propagate(
            clauses, unassigned
        )
        if conflict:
            return Fraction(0)

        # All clauses satisfied after propagation
        if not up_cls:
            return up_weight * self._free_count(up_unass)

        # Separate free variables (not in any remaining clause) from active ones
        clause_vars = set()
        for clause in up_cls:
            for lit in clause:
                clause_vars.add(abs(lit))
        active = up_unass & clause_vars
        free = up_unass - clause_vars
        free_factor = self._free_count(free)

        # Cache lookup (key = normalized remaining clauses;
        # free vars are excluded so cache is independent of them)
        cache_key = tuple(sorted(up_cls))
        if cache_key in self.cache:
            return up_weight * free_factor * self.cache[cache_key]

        # Component decomposition
        components = self.find_components(up_cls, active)

        if len(components) > 1:
            result = Fraction(1)
            for comp_vars, comp_clauses in components:
                result *= self.count(comp_clauses, comp_vars)

            self.cache[cache_key] = result
            return up_weight * free_factor * result

        # Single component: branch on the most frequent variable
        var_freq = defaultdict(int)
        for clause in up_cls:
            for lit in clause:
                v = abs(lit)
                if v in active:
                    var_freq[v] += 1

        branch_var = max(var_freq, key=var_freq.get)

        result = Fraction(0)
        for val in [True, False]:
            lit = branch_var if val else -branch_var
            w = self.lit_weight(lit)
            new_cls = self.simplify(up_cls, branch_var, val)
            if new_cls is None:
                continue  # conflict on this branch
            new_unass = active - {branch_var}
            result += w * self.count(new_cls, new_unass)

        self.cache[cache_key] = result
        return up_weight * free_factor * result

    def solve(self):
        """Compute the exact model count."""
        all_vars = set(range(1, self.num_vars + 1))
        return self.count(list(self.clauses), all_vars)


def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <cnf_file>", file=sys.stderr)
        sys.exit(1)

    num_vars, clauses, mode, weights = parse_dimacs(sys.argv[1])
    counter = ModelCounter(num_vars, clauses, mode, weights)
    result = counter.solve()

    is_weighted = mode in ('wmc', 'pwmc', 'wpmc')

    if result == 0:
        print("s UNSATISFIABLE")
    else:
        print("s SATISFIABLE")

    if is_weighted:
        print("c s type wmc")
        print(f"c s exact arb float {float(result)}")
        frac = Fraction(result)
        print(f"c s exact rational {frac.numerator}/{frac.denominator}")
    else:
        count_int = int(result)
        print("c s type mc")
        if count_int > 0:
            print(f"c s log10-estimate {math.log10(count_int)}")
        print(f"c s exact arb int {count_int}")


if __name__ == '__main__':
    main()
